### ISSUE TITLE
Initial executable memory allocation on ARM64 Linux is not right

### ISSUE THREAD

**Issue by janvorli** [dotnet/runtime#65900]
I've found that on ARM64 Linux, we use a lot of jumpstubs. I have investigated why that happens and I have found that the initial preallocation of the executable memory range close to libcoreclr.so is based on x64 pc relative jump range, which is 16 times larger than on arm64. Moreover, the size of the libcoreclr.so that we use there is heavily overestimated. We use 100 MB, which is an approx size of libcoreclr.so with debugging symbols. But those symbols are not part of the in-memory image and the real size is about 10 times slower.
This all makes getting executable memory close to libcoreclr.so extremely unlikely, thus inducing the large number of jumpstubs.

I have experimentally scaled down the range and coreclr size and I was able to get 0 jumpstubs in cases where we've used more than 2000 before.

**Comment by janvorli** [comment]
cc: @EgorBo

**Comment by EgorBo** [comment]
@janvorli heh, yeah we've investigated it in https://github.com/dotnet/runtime/pull/63842 and https://github.com/dotnet/runtime/issues/62302

> I have experimentally scaled down the range

Long story short, it does improve perf but it also means that a typical large app will quickly fill that initial heap and will have to allocate a new one (most likely too far from the initial one and coreclr.so/dylib)

**Comment by jkotas** [comment]
> There are two solutions as we see them:

There is also 3rd solution: Create local copies of indirections cells and small hot helpers.

The typical call sequence once @janvorli refactoring goes in is going to look like this:
```
movz #x0, #0x....
movk #x0, #0x.... LSL #16
movk #x0, #0x.... LSL #32
ldr x0, [x0]
blr x0
```
The idea would be to turn it into this:
```
ldr x0, [pc + 0x....]
blr x0
```

**Comment by jkotas** [comment]
> Make sure coreclr, corelib, and, say 512mb of user's code lay next to each other

I do not think 512mb would be small enough for arm64. Also, designs like this tend to be in conflict with https://github.com/dotnet/designs/blob/main/accepted/2021/runtime-security-mitigations.md#address-space-layout-randomization-aslr

**Comment by EgorBo** [comment]
> Also, designs like this tend to be in conflict with https://github.com/dotnet/designs/blob/main/accepted/2021/runtime-security-mitigations.md#address-space-layout-randomization-aslr

Right, but looks like we manage to avoid ASLR effects somehow on x86 arch in our memory manager and always manage to reserve a space next to coreclr 😐

**Comment by janvorli** [comment]
@EgorBo I am sorry for the delayed response. Regarding making the GetRelocHint super accurate, the issue is that once we are out of the pre-reserved memory the OS decides where to reserve the next block of memory. It can be random for SELinux, it can be randomized by other stuff allocated / loaded in the memory. If we don't know the length of the jitted code in advance, it is not possible to guess whether it will fit into the remainder of the currently reserved block or not. And there can be multiple threads competing for the memory from that block. Pre-allocation using some guesstimate seems like a possible solution.

As for putting code close to the libcoreclr.so, since we cannot control where the libcoreclr.so is loaded, the current mechanism that tries to preallocate memory at PAL initialization seems to be the best effort to do that. It seems that with R2R enabled, there is a high chance that S.P.C. R2R code loaded from the file will end up close to libcoreclr.so. But with ASLR enabled, its success is kind of random. And with SELinux enabled, the mmap doesn't honor the requested base address and always uses a random one. Although in this case, we could say that higher security has some perf cost.

As mentioned somewhere in the related issues here, we could possibly try to consider the preallocated space "precious" and use it only for some allocations. E.g. for allocating S.P.C. R2R and JITted code and related stubs, maybe for some more runtime assemblies from there, but not for upper layer's code.

As for @jkotas idea to create local copies of indirections cells with my upcoming refactorings, I am not sure how we would maintain coherency of these copies with the slot that's part of the precode. The fixup precode slot initially points to the fixup code, then it changes to point to the target code and it also has ResetTargetInterlocked method that can reset the slot to point to the fixup code again.

**Comment by jkotas** [comment]
> I am not sure how we would maintain coherency of these copies with the slot that's part of the precode.

We do have copies of slot values today. `MethodDescBackpatchInfoTracker` keeps track of them and makes sure to patch all copies as needed.

**Comment by janvorli** [comment]
This was fixed by #70707
